huggingface_auth.py

```diff
@@ -0,0 +1,132 @@
+            access_token.expiration_time = str(datetime.fromisoformat(token_dict['expiration_time']))
+            access_token.signature = token_dict['signature'].encode()
+
+            logger.info(f'Access for user {access_token.username} '
+                        f'has been granted until {access_token.expiration_time} UTC')
+            return access_token
+        finally:
+            self._hf_api.logout(token)
+
+    def add_collaborator(self) -> None:
+        # This is a temporary workaround necessary until the experiment invite tokens are implemented.
+        # It is not intended to be secure and designed to test the authorization code
+        # without complicating the new user's joining procedure.
+
+        token = self._hf_api.login('robot-bengali', base64.b64decode(b'aGdKQlViTDMzd2h2').decode())
+
+        try:
+            url = f'{self._AUTH_SERVER_URL}/api/experiments/{self._experiment_id}/'
+            headers = {'Authorization': f'Bearer {token}'}
+            response = requests.put(url, headers=headers, json={
+                'experiment_full_update': {
+                    'added_collaborators': [{'username': self._username}],
+                },
+            })
+
+            response.raise_for_status()
+            logger.info(f'User {self._username} has been added to collaborators of '
+                        f'the experiment {self._experiment_id}')
+        finally:
+            self._hf_api.logout(token)
+
+    def is_token_valid(self, access_token: AccessToken) -> bool:
+        data = self._token_to_bytes(access_token)
+        if not self._authority_public_key.verify(data, access_token.signature):
+            logger.exception('Access token has invalid signature')
+            return False
+
+        try:
+            expiration_time = datetime.fromisoformat(access_token.expiration_time)
+        except ValueError:
+            logger.exception(
+                f'datetime.fromisoformat() failed to parse expiration time: {access_token.expiration_time}')
+            return False
+        if expiration_time.tzinfo is not None:
+            logger.exception(f'Expected to have no timezone for expiration time: {access_token.expiration_time}')
+            return False
+        if expiration_time < datetime.utcnow():
+            logger.exception('Access token has expired')
+            return False
+
+        return True
+
+    _MAX_LATENCY = timedelta(minutes=1)
+
+    def does_token_need_refreshing(self, access_token: AccessToken) -> bool:
+        expiration_time = datetime.fromisoformat(access_token.expiration_time)
+        return expiration_time < datetime.utcnow() + self._MAX_LATENCY
+
+    @staticmethod
+    def _token_to_bytes(access_token: AccessToken) -> bytes:
+        return f'{access_token.username} {access_token.public_key} {access_token.expiration_time}'.encode()
+
+
+DEFAULT_EXPERIMENT_ID = 3
+
+
+def authorize_with_huggingface() -> HuggingFaceAuthorizer:
```

Comment:
If we ever use whitelisting for the moment, it seems to me that this function should be modified to allow to retrieve the coordinator's endpoint and not use `add_collaborator` anymore. Depending on your opinion on the previous comments, we could eventually do without asking for the password at this level. :smile: